src/tasks/LibraryBuilder/Templates/autoinit.c

```diff
@@ -0,0 +1,116 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+
+#include <mono/jit/jit.h>
+#include <mono/jit/mono-private-unstable.h>
+#include <mono/metadata/assembly.h>
+
+#ifdef HOST_ANDROID
+#include <android/log.h>
+
+#define LOG_INFO(fmt, ...) __android_log_print(ANDROID_LOG_INFO, "MONO_SELF_CONTAINED_LIBRARY", fmt, ##__VA_ARGS__)
+#define LOG_ERROR(fmt, ...) __android_log_print(ANDROID_LOG_ERROR, "MONO_SELF_CONTAINED_LIBRARY", fmt, ##__VA_ARGS__)
+#else
+#include <os/log.h>
+
+#define LOG_INFO(fmt, ...) os_log_info (OS_LOG_DEFAULT, fmt, ##__VA_ARGS__)
+#define LOG_ERROR(fmt, ...) os_log_error (OS_LOG_DEFAULT, fmt, ##__VA_ARGS__)
+#endif
+
+static const char *bundle_path;
+
+void register_aot_modules (void);
+void load_assemblies_with_exported_symbols ();
+typedef void (*MonoRuntimeInitCallback) (void);
+void mono_set_runtime_init_callback (MonoRuntimeInitCallback callback);
+
+static void
+cleanup_runtime_config (MonovmRuntimeConfigArguments *args, void *user_data)
+{
+    free (args);
+    free (user_data);
+}
+
+static void
+initialize_runtimeconfig ()
+{
+    char *file_name = "runtimeconfig.bin";
+    size_t str_len = sizeof (char) * (strlen (bundle_path) + strlen (file_name) + 2); // +1 "/", +1 null-terminating char
+    char *file_path = (char *)malloc (str_len);
+    if (!file_path) {
+        LOG_ERROR ("Out of memory.\n");
+        abort ();
+    }
+
+    int num_char = snprintf (file_path, str_len, "%s/%s", bundle_path, file_name);
+    struct stat buffer;
+
+    if (stat (file_path, &buffer) == 0) {
+        MonovmRuntimeConfigArguments *arg = (MonovmRuntimeConfigArguments *)malloc (sizeof (MonovmRuntimeConfigArguments));
+        arg->kind = 0;
+        arg->runtimeconfig.name.path = file_path;
+        monovm_runtimeconfig_initialize (arg, cleanup_runtime_config, file_path);
+    } else {
+        LOG_INFO ("Could not stat file '%s'. Runtime configuration properties not initialized.\n", file_path);
+        free (file_path);
+    }
+}
+
+static void
+initialize_appctx_env_variables ()
+{
+    const char *appctx_keys[2], *appctx_values[2];
+
+    appctx_keys[0] = "RUNTIME_IDENTIFIER";
+    appctx_values[0] = "%RUNTIME_IDENTIFIER%";
+
+    appctx_keys[1] = "APP_CONTEXT_BASE_DIRECTORY";
+    appctx_values[1] = bundle_path;
+
+    monovm_initialize(2, appctx_keys, appctx_values);
+}
+
+void
+runtime_init_callback ()
+{
+    const char *assemblies_path = strdup(getenv("%ASSEMBLIES_LOCATION%"));
```

Comment:
getenv returns NULL if variable is not found.

Reply:
Good catch, 
changed to 
```
    const char *assemblies_path = getenv("%ASSEMBLIES_LOCATION%");
    if (assemblies_path && assemblies_path[0] != '\0')
        bundle_path = strdup (assemblies_path);
    else
        bundle_path = strdup ("./");
    if (!bundle_path)
        LOG_ERROR ("Out of memory.\n");
```

and freeing `bundle_path` at the end of the init callback instead of `assemblies_path` since [It isn't safe to modify the value of the environment variable using the returned pointer.](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/getenv-wgetenv?view=msvc-170)

Reply:
After seeing https://github.com/dotnet/runtime/pull/83050#discussion_r1136792693
```
    bundle_path = getenv("%ASSEMBLIES_LOCATION%");
    if (assemblies_path && assemblies_path[0] != '\0')
        bundle_path = "./";
```
and no need to free afterwards.